scripts/build/chroot_generator.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package build
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+func init() {
+	archiveGenerators["chroot"] = chrootGenerator{}
+}
+
+type chrootGenerator struct {
+}
+
+// The chroot generator dumps the rootfs tree into a directory appropriate for
+// running under a chroot and namespaces.
+func (g chrootGenerator) generate(config Config, files []file) error {
+	// Since files is sorted, we can guarantee that the directories are created
+	// before their children.
+	for _, f := range files {
+		f.path = filepath.Join(config.OutputPath, f.path)
+		if err := createFile(f); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Create an actual file from the file struct.
+func createFile(f file) error {
+	// TODO: set uid and gid
+	mode := fmt.Sprintf("%03o", f.mode&os.ModePerm)
+	major := fmt.Sprint(major(f.rdev))
+	minor := fmt.Sprint(minor(f.rdev))
+
+	// Special file types
+	// I'm sure many of these commands can be done through syscalls rather than
+	// forking, however we need to perform the priviledge escalation via sudo
+	// (asking for their password).
+	switch f.mode & (os.ModeType | os.ModeCharDevice) {
+	case os.ModeDir:
+		return exec.Command("sudo", "mkdir", "--mode="+mode, f.path).Run()
+	case os.ModeDevice:
+		return exec.Command("sudo", "mknod", "--mode="+mode, f.path, "b", major, minor).Run()
+	case os.ModeDevice | os.ModeCharDevice:
+		return exec.Command("sudo", "mknod", "--mode="+mode, f.path, "c", major, minor).Run()
+	case os.ModeNamedPipe:
+		return exec.Command("sudo", "mknod", "--mode="+mode, f.path, "p").Run()
+	case os.ModeSymlink:
+		if err := exec.Command("sudo", "ln", "-s", string(f.data), f.path).Run(); err != nil {
+			return err
+		}
+		return exec.Command("sudo", "chmod", mode, f.path).Run()
+	case os.ModeSocket:
+		return errors.New("making sockets not supported yet") // TODO
+	}
+
+	// Regular files
+	cmd := exec.Command("sudo", "tee", f.path)
+	cmd.Stdin = bytes.NewBuffer(f.data)
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+	return exec.Command("sudo", "chmod", mode, f.path).Run()
+}
+
+// Run the rootfs under a chroot jail.
+func (g chrootGenerator) run(config Config) error {
```

Comment:
Is there a solution to the `--propagation=private` problem? It is pretty important for the mount namespace to be useful at all.